developer/frameworks/articles/290-configuration/09-fully-custom-system-settings-ui.markdown

```diff
@@ -0,0 +1,292 @@
+
+    request.setAttribute(
+        LanguageTemplateConfigurationDisplayContext.class.getName(),
+        languageTemplateConfigurationDisplayContext);
+
+    _jspRenderer.renderJSP(
+        _servletContext, request, response,
+        "/configuration/site_navigation_language_web_template.jsp");
+}
+
+```
+
+Specify the required service references at the bottom of the class. Be careful
+to target the proper servlet context, passing the `bundle-SymbolicName` of the
+module (found in its `bnd.bnd` file) into the `osgi.web.symbolicname` property
+of the reference target:
+
+```java
+
+@Reference
+private DDMTemplateLocalService _ddmTemplateLocalService;
+
+@Reference
+private GroupLocalService _groupLocalService;
+
+@Reference
+private JSPRenderer _jspRenderer;
+
+@Reference
+private Portal _portal;
+
+@Reference(
+    target = "(osgi.web.symbolicname=com.liferay.site.navigation.language.web)",
+    unbind = "-"
+)
+private ServletContext _servletContext;
+
+```
+
+Once the configuration form renderer is implemented, you can write the JSP
+markup for the form.
+
+## Writing the JSP Markup [](id=writing-the-jsp-markup)
+
+Now write the JSP:
+
+```jsp
+
+    <%@ include file="/init.jsp" %>
+
+    <%
+    LanguageTemplateConfigurationDisplayContext
+        languageTemplateConfigurationDisplayContext = (LanguageTemplateConfigurationDisplayContext)request.getAttribute(LanguageTemplateConfigurationDisplayContext.class.getName());
+
+Admin: Instance Settings    String currentTemplateName = languageTemplateConfigurationDisplayContext.getCurrentTemplateName();
+    %>
+
+    <aui:select label="<%= HtmlUtil.escape(languageTemplateConfigurationDisplayContext.getFieldLabel()) %>" name="ddmTemplateKey" value="<%= currentTemplateName %>">
+
+        <%
+        for (String[] templateValue : languageTemplateConfigurationDisplayContext.getTemplateValues()) {
+        %>
+
+            <aui:option label="<%= templateValue[1] %>" selected="<%= currentTemplateName.equals(templateValue[0]) %>" value="<%= templateValue[0] %>" />
+
+        <%
+        }
+        %>
+
+    </aui:select>
+
+```
+
+The opening scriptlet gets the display context object from the request so that
+all its getters are invoked whenever information from the back-end is required.
+Right away, the `getCurrentTemplateName` method is called, since the current
+template name is needed for the first option's `ddmTemplateKey` display value as
+soon as the form is rendered. This happens in the `<aui:select>` tag. There's
+just a bit of logic used to create an option for each of the available DDM
+templates that can be chosen.
+
+So what does this example look like when all is said and done?
+
+![Figure 2: A select list provides a more user friendly configuration experience than a text field.](../../../images/sys-settings-lang-template-custom.png)
+
+Some configuration UIs require tweaking with a `ConfigurationFormRenderer`.
+This tutorial shows a particularly good example. Administrators encountering
```

Comment:
Are we saying that the example itself is really good, or that the `ConfigurationFormRenderer` tool is particularly good for this use-case? It seems like the latter is more appropriate, even though I think the example code I wrote is decent too ;)